gpt_engineer/core/default/steps.py

```diff
@@ -0,0 +1,183 @@
+from gpt_engineer.core.code import Code
+from gpt_engineer.core.ai import AI
+from gpt_engineer.core.chat_to_files import (
+    parse_chat,
+    overwrite_code_with_edits,
+    parse_edits,
+)
+from gpt_engineer.core.default.paths import (
+    ENTRYPOINT_FILE,
+    CODE_GEN_LOG_FILE,
+    ENTRYPOINT_LOG_FILE,
+    IMPROVE_LOG_FILE,
+    PREPROMPTS_PATH,
+)
+from gpt_engineer.core.default.constants import MAX_EDIT_REFINEMENT_STEPS
+from gpt_engineer.core.default.on_disk_repository import OnDiskRepository
+from gpt_engineer.core.base_repository import BaseRepository
+from gpt_engineer.core.base_execution_env import BaseExecutionEnv
+
+from typing import Union, MutableMapping, List
+from pathlib import Path
+from langchain.schema import HumanMessage, SystemMessage
+import inspect
+import re
+from termcolor import colored
+from gpt_engineer.core.preprompts_holder import PrepromptsHolder
+
+
+def curr_fn() -> str:
+    return inspect.stack()[1].function
+
+
+def setup_sys_prompt(preprompts: MutableMapping[Union[str, Path], str]) -> str:
+    return (
+        preprompts["roadmap"]
+        + preprompts["generate"].replace("FILE_FORMAT", preprompts["file_format"])
+        + "\nUseful to know:\n"
+        + preprompts["philosophy"]
+    )
+
+
+def gen_code(
+    ai: AI, prompt: str, memory: BaseRepository, preprompts_holder: PrepromptsHolder
+) -> Code:
+    preprompts = preprompts_holder.get_preprompts()
+    messages = ai.start(setup_sys_prompt(preprompts), prompt, step_name=curr_fn())
+    chat = messages[-1].content.strip()
+    memory[CODE_GEN_LOG_FILE] = chat
+    files = parse_chat(chat)
+    code = Code({key: val for key, val in files})
```

Comment:
minor point: this is identical to `Code(**files)`

Reply:
Hmm, I don't get that to work for files, which is List[Tuple[str, str]] .

>>> a = [("z","x"), ("y","w")]
>>> b = dict(**a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: dict() argument after ** must be a mapping, not list

Note that Code inherits dict.